src/mlpack/core/tree/vantage_point_tree/vantage_point_tree_impl.hpp

```diff
@@ -0,0 +1,921 @@
+         typename MatType,
+         template<typename BoundMetricType, typename...> class BoundType,
+         template<typename SplitBoundType, typename SplitMatType>
+             class SplitType>
+inline size_t VantagePointTree<MetricType, StatisticType, MatType, BoundType,
+                              SplitType>::Point(const size_t index) const
+{
+  return (begin + index);
+}
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename BoundMetricType, typename...> class BoundType,
+         template<typename SplitBoundType, typename SplitMatType>
+             class SplitType>
+void VantagePointTree<MetricType, StatisticType, MatType, BoundType, SplitType>::
+    SplitNode(const size_t maxLeafSize,
+              SplitType<BoundType<MetricType>, MatType>& splitter)
+{
+  // We need to expand the bounds of this node properly.
+  if (parent)
+  {
+    bound.Center() = parent->firstPointIsCentroid ?
+			dataset->col(parent->begin + 1) : dataset->col(parent->begin);
+    bound.OuterRadius() = 0;
+    bound.InnerRadius() = std::numeric_limits<ElemType>::max();
+  }
+
+  if (count > 0)
+    bound |= dataset->cols(begin, begin + count - 1);
+
+  VantagePointTree* tree = this;
+
+  while (tree->Parent() != NULL)
+  {
+    tree->Parent()->Bound() |= tree->Bound();
+	tree->Parent()->furthestDescendantDistance = 0.5 *
+			tree->Parent()->Bound().Diameter();
+    tree = tree->Parent();
+  }
+  // Calculate the furthest descendant distance.
+  furthestDescendantDistance = 0.5 * bound.Diameter();
+
+  // Now, check if we need to split at all.
+  if (count <= maxLeafSize)
+    return; // We can't split this.
+
+  // splitCol denotes the two partitions of the dataset after the split. The
+  // points on its left go to the left child and the others go to the right
+  // child.
+  size_t splitCol;
+
+  // Split the node. The elements of 'data' are reordered by the splitting
+  // algorithm. This function call updates splitCol.
+
+  size_t splitBegin = begin;
+  size_t splitCount = count;
+
+  if (IsFirstPointCentroid())
+  {
+    splitBegin = begin + 1;
+    splitCount = count - 1;
+  }
+
+  const bool split = splitter.SplitNode(bound, *dataset, splitBegin, splitCount,
+      splitCol);
+
+  // The node may not be always split. For instance, if all the points are the
+  // same, we can't split them.
+  if (!split)
+    return;
+
+  // Now that we know the split column, we will recursively split the children
+  // by calling their constructors (which perform this splitting process).
+  left = new VantagePointTree(this, splitBegin, splitCol - splitBegin, splitter,
+      maxLeafSize, true);
+  right = new VantagePointTree(this, splitCol, splitBegin + splitCount - splitCol,
+      splitter, maxLeafSize, false);
+
+  // Calculate parent distances for those two nodes.
+  arma::vec center, leftCenter, rightCenter;
+  Center(center);
+  left->Center(leftCenter);
+  right->Center(rightCenter);
```

Comment:
Shouldn't the parent distance be identical for both children?  (I think it will just be the distance between this node's vantage point and the left child's vantage point)